### PR TITLE
Fix n3 parser exponent syntax of floats with leading dot.

### LINKED ISSUE #999: Turtle parser fails with leading dot in decimal value
Input file `test.ttl`:
```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> 

<http://qudt.org/vocab/unit/MilliM-PER-YR>
  a <http://qudt.org/schema/qudt/Unit> ;
  <http://qudt.org/schema/qudt/conversionMultiplier> .171e-11 ;
  <http://qudt.org/schema/qudt/conversionOffset> 0e+00 ;
  <http://qudt.org/schema/qudt/description> "0.001-fold of the SI base unit metre divided by the unit year" ;
  <http://qudt.org/schema/qudt/hasQuantityKind> <http://qudt.org/vocab/quantitykind/Velocity> ;
  <http://qudt.org/schema/qudt/iec61360Code> "0112/2///62720#UAA868" ;
  <http://qudt.org/schema/qudt/uneceCommonCode> "H66" ;
  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
  rdfs:isDefinedBy <http://qudt.org/vocab/unit> ;
  rdfs:label "MilliM PER YR" ;
  <http://www.w3.org/2004/02/skos/core#prefLabel> "millimetre per year" ;
.
```
Traceback:
```
>>> from rdflib import Graph
INFO:rdflib:RDFLib Version: 4.2.2
>>> g = Graph().parse('test.ttl', format='turtle')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/rdflib/graph.py", line 1043, in parse
    parser.parse(source, self, **args)
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 1870, in parse
    p.loadStream(source.getByteStream())
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 434, in loadStream
    return self.loadBuf(stream.read())    # Not ideal
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 440, in loadBuf
    self.feed(buf)
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 466, in feed
    i = self.directiveOrStatement(s, j)
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 489, in directiveOrStatement
    return self.checkDot(argstr, j)
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 1159, in checkDot
    "expected '.' or '}' or ']' at end of statement")
  File "/usr/local/lib/python3.7/site-packages/rdflib/plugins/parsers/notation3.py", line 1615, in BadSyntax
    raise BadSyntax(self._thisDoc, self.lines, argstr, i, msg)
rdflib.plugins.parsers.notation3.BadSyntax: at line 5 of <>:
Bad syntax (expected '.' or '}' or ']' at end of statement) at ^ in:
"...b' ;\n  <http://qudt.org/schema/qudt/conversionMultiplier> .171'^b'e-11 ;\n  <http://qudt.org/schema/qudt/conversionOffset> 0e+0'..."
>>> exit()
```
The parser is having issue with this line in the input file: `<http://qudt.org/schema/qudt/conversionMultiplier> .171e-11 ;`

Not very familiar with the code but probably some additional checks are required in `directiveOrStatement()` or `checkDot()`.

### PR DESCRIPTION
**Pull request by white-gecko**
Fix #999 .

### CODE REVIEW

File: rdflib/plugins/parsers/notation3.py [hunk 1]
```diff
@@ -349,7 +349,7 @@ def unicodeExpand(m):
 signed_integer = re.compile(r'[-+]?[0-9]+')      # integer
 integer_syntax = re.compile(r'[-+]?[0-9]+')
 decimal_syntax = re.compile(r'[-+]?[0-9]*\.[0-9]+')
-exponent_syntax = re.compile(r'[-+]?(?:[0-9]+\.[0-9]*(?:e|E)[-+]?[0-9]+|'+
+exponent_syntax = re.compile(r'[-+]?(?:[0-9]*\.[0-9]*(?:e|E)[-+]?[0-9]+|'+
                              r'\.[0-9](?:e|E)[-+]?[0-9]+|'+
```
Comment (effigies) [inline rdflib/plugins/parsers/notation3.py:353]: Your fix will now match ".e1".

What it looks like this was trying to do is allow ".1e1", "1.e1" and "1e1", which means you need `[0-9]*\.[0-9]+`, `[0-9]+\.[0-9]*` and `[0-9]+`. The problem is line 2 is missing a `+`:

```suggestion
exponent_syntax = re.compile(r'[-+]?(?:[0-9]+\.[0-9]*(?:e|E)[-+]?[0-9]+|'+
                             r'\.[0-9]+(?:e|E)[-+]?[0-9]+|'+
```

**Comment by effigies** [comment]
Hopefully I'm not being annoying, but as I was looking at this I wanted to add some tests that should have caught this problem earlier. Made a proposal at white-gecko#1.

**Comment by white-gecko** [comment]
@effigies everything fine. We are happy about all good contributions.

File: test/test_n3.py [hunk 2]
```diff
@@ -226,5 +252,24 @@ def testEmptyPrefix(self):
...
+    def testExponents(self):
+        signs = ("", "+", "-")
+        mantissas = ("1", "1.", ".1",
+                     "12", "12.", "1.2", ".12",
+                     "123", "123.", "12.3", "1.23", ".123")
+        es = "eE"
+        exps = ("1", "12", "+1", "-1", "+12", "-12")
+        for parts in itertools.product(signs, mantissas, es, exps):
+            expstring = "".join(parts)
+            self.assertRegex(expstring, exponent_syntax)
```
Comment (effigies) [inline test/test_n3.py:265]:
```suggestion
            self.assertIsNotNone(exponent_syntax.match(expstring))
```

File: test/test_n3.py [hunk 3]
```diff
@@ -226,5 +252,24 @@ def testEmptyPrefix(self):
...
+
+    def testInvalidExponents(self):
+        # Add test cases as needed
+        invalid = (".e1",)
+        for expstring in invalid:
+            self.assertNotRegex(expstring, exponent_syntax)
```
Comment (effigies) [inline test/test_n3.py:271]:
```suggestion
            self.assertIsNone(exponent_syntax(expstring))
```

File: test/test_n3.py (hunk 2 above)
Comment (white-gecko) [reply test/test_n3.py]: I used `assetTrue` and `assertFalse` which are effectively doing the same.

**Comment by gweis** [comment]
@effigies nice find with the regex. Thanks.

**Pull request merged**